File: pipeline/scoring.py
"""
Python mirror of web/scoring.js.

Kept in lockstep by pipeline/test_scoring.py, which runs both over random
protein pairs and asserts identical output. If you change one, change the
other and run that test — the simulator's verdict on game difficulty is
only meaningful if it scores exactly the way the game does.
"""

CORRECT, PARTIAL, WRONG = "correct", "partial", "wrong"
UP, DOWN = "↑", "↓"

COLUMNS = ["len", "con", "loc", "fn", "pw", "chr", "dis"]
COLUMN_LABELS = ["Length", "Conservation", "Localization", "Function",
                 "Pathway", "Chromosome", "Disease"]
LENGTH_CLOSE = 0.10


def chrom_order(c):
    """Autosomes 1-22, then X, Y, MT, so the arrow spans the karyotype."""
    if c is None or c == "":
        return None
    if c == "X":
        return 23
    if c == "Y":
        return 24
    if c == "MT":
        return 25
    try:
        return int(c)
    except (TypeError, ValueError):
        return None


def cmp_numeric(g, t, close_fraction):
    if g is None or t is None:
        return (WRONG, None)
    if g == t:
        return (CORRECT, None)
    arrow = UP if t > g else DOWN
    if close_fraction:
        if abs(t - g) / max(t, 1) <= close_fraction:
            return (PARTIAL, arrow)
    return (WRONG, arrow)


def cmp_ordinal(g_key, t_key, ladder_rank):
    g = ladder_rank.get(g_key)
    t = ladder_rank.get(t_key)
    if g is None or t is None:
        return (WRONG, None)
    if g == t:
        return (CORRECT, None)
    arrow = UP if t > g else DOWN
    return (PARTIAL if abs(t - g) == 1 else WRONG, arrow)


def cmp_set(g_arr, t_arr):
    g = list(g_arr or [])
    t = list(t_arr or [])
    if not g and not t:
        return (CORRECT, None)
    ts = set(t)
    if len(g) == len(t) and all(v in ts for v in g):
        return (CORRECT, None)
    return (PARTIAL if any(v in ts for v in g) else WRONG, None)


def cmp_exact(g, t):
    return (CORRECT if g == t else WRONG, None)


def cmp_grouped(g, t, groups):
    """Exact class green; a sibling class in the same family amber."""
    if g == t:
        return (CORRECT, None)
    gg = groups.get(g)
    if gg is not None and gg == groups.get(t):
        return (PARTIAL, None)
    return (WRONG, None)


def compare(guess, target, meta):
    """
    Seven (state, arrow) pairs, in board order.

    `meta` holds the lookup tables: {"ladderRank": ..., "functionGroups":
    ...}, both emitted into proteins.json at build time.
    """
    ladder_rank = meta.get("ladderRank", {})
    groups = meta.get("functionGroups", {})
    return [
        cmp_numeric(guess.get("len"), target.get("len"), LENGTH_CLOSE),
        cmp_ordinal(guess.get("con"), target.get("con"), ladder_rank),
        cmp_set(guess.get("loc"), target.get("loc")),
        cmp_grouped(guess.get("fn"), target.get("fn"), groups),
        cmp_set(guess.get("pw"), target.get("pw")),
        cmp_numeric(chrom_order(guess.get("chr")),
                    chrom_order(target.get("chr")), 0),
        cmp_exact(guess.get("dis"), target.get("dis")),
    ]


def signature(guess, target, meta):
    """
    A hashable summary of one guess's feedback. Two candidate answers with
    the same signature are indistinguishable to the player after that
    guess, which is what makes this the right key for pruning.
    """
    return "|".join(s + (a or "")
                    for s, a in compare(guess, target, meta))


def meta_from(data):
    """Build the lookup tables the rules need from a loaded proteins.json."""
    return {
        "ladderRank": {r["key"]: r["rank"] for r in data["ladder"]},
        "functionGroups": data.get("functionGroups", {}),
    }

File: web/scoring.js
/* -------------------------------------------------------------------
   Proteindle scoring — the single definition of what a guess tells you.

   Loaded as a plain <script> by the game and require()d by the Node-side
   cross-check test. pipeline/scoring.py mirrors it exactly, and
   pipeline/test_scoring.py runs both over random pairs and asserts they
   agree — because the simulator's conclusions about game difficulty are
   worthless if it scores differently from the game itself.

   Every comparison returns {state, arrow}:
     state — 'correct' | 'partial' | 'wrong'
     arrow — '↑' (answer is higher) | '↓' (lower) | null
   ------------------------------------------------------------------- */

(function (root, factory) {
  if (typeof module === 'object' && module.exports) module.exports = factory();
  else root.ProteindleScoring = factory();
}(typeof self !== 'undefined' ? self : this, function () {
  'use strict';

  // Sex chromosomes and the mitochondrial genome sort after the autosomes
  // so the up/down arrow stays meaningful across the whole karyotype.
  function chromOrder(c) {
    if (c === null || c === undefined || c === '') return null;
    if (c === 'X') return 23;
    if (c === 'Y') return 24;
    if (c === 'MT') return 25;
    const n = parseInt(c, 10);
    return Number.isNaN(n) ? null : n;
  }

  function cmpNumeric(guessVal, targetVal, closeFraction) {
    if (guessVal === null || guessVal === undefined ||
        targetVal === null || targetVal === undefined) {
      return { state: 'wrong', arrow: null };
    }
    if (guessVal === targetVal) return { state: 'correct', arrow: null };
    const arrow = targetVal > guessVal ? '↑' : '↓';
    if (closeFraction) {
      const diff = Math.abs(targetVal - guessVal) / Math.max(targetVal, 1);
      if (diff <= closeFraction) return { state: 'partial', arrow: arrow };
    }
    return { state: 'wrong', arrow: arrow };
  }

  function cmpOrdinal(guessKey, targetKey, ladderRank) {
    const g = ladderRank[guessKey];
    const t = ladderRank[targetKey];
    if (g === undefined || t === undefined) {
      return { state: 'wrong', arrow: null };
    }
    if (g === t) return { state: 'correct', arrow: null };
    const arrow = t > g ? '↑' : '↓';
    // One rung apart is genuinely close on a seven-rung ladder.
    const near = Math.abs(t - g) === 1;
    return { state: near ? 'partial' : 'wrong', arrow: arrow };
  }

  function cmpSet(guessArr, targetArr) {
    const g = guessArr || [];
    const t = targetArr || [];
    if (g.length === 0 && t.length === 0) {
      return { state: 'correct', arrow: null };
    }
    const ts = new Set(t);
    if (g.length === t.length && g.every(function (v) { return ts.has(v); })) {
      return { state: 'correct', arrow: null };
    }
    const overlap = g.some(function (v) { return ts.has(v); });
    return { state: overlap ? 'partial' : 'wrong', arrow: null };
  }

  function cmpExact(guessVal, targetVal) {
    return {
      state: guessVal === targetVal ? 'correct' : 'wrong',
      arrow: null,
    };
  }

  // Exact class is green; a different class in the same family is amber.
  // "Kinase" against "Protease" both being enzymes is a real, useful hint.
  function cmpGrouped(guessVal, targetVal, groups) {
    if (guessVal === targetVal) return { state: 'correct', arrow: null };
    const g = groups[guessVal];
    const t = groups[targetVal];
    if (g !== undefined && g === t) return { state: 'partial', arrow: null };
    return { state: 'wrong', arrow: null };
  }

  // Column order here is the column order on the board.
  const COLUMNS = ['len', 'con', 'loc', 'fn', 'pw', 'chr', 'dis'];

  // Length tolerance. Proteins vary over three orders of magnitude, so a
  // proportional band reads better than an absolute one: 10% of a 100 aa
  // peptide is 10 aa, 10% of a 2000 aa monster is 200.
  const LENGTH_CLOSE = 0.10;

  /* `meta` carries the lookup tables the rules need: {ladderRank,
     functionGroups}. Both are emitted into proteins.json at build time so
     the rules here stay data-driven rather than hard-coding biology. */
  function compare(guess, target, meta) {
    const ladderRank = (meta && meta.ladderRank) || {};
    const groups = (meta && meta.functionGroups) || {};
    return [
      cmpNumeric(guess.len, target.len, LENGTH_CLOSE),
      cmpOrdinal(guess.con, target.con, ladderRank),
      cmpSet(guess.loc, target.loc),
      cmpGrouped(guess.fn, target.fn, groups),
      cmpSet(guess.pw, target.pw),
      cmpNumeric(chromOrder(guess.chr), chromOrder(target.chr), 0),
      cmpExact(guess.dis, target.dis),
    ];
  }

  /* A compact, comparable signature of one guess's feedback. Two answers
     that produce the same signature are indistinguishable to the player
     after that guess — which is exactly what the solver needs to prune
     candidates, and what the entropy report needs to count. */
  function signature(guess, target, meta) {
    return compare(guess, target, meta)
      .map(function (r) { return r.state + (r.arrow || ''); })
      .join('|');
  }

  return {
    chromOrder: chromOrder,
    cmpNumeric: cmpNumeric,
    cmpOrdinal: cmpOrdinal,
    cmpSet: cmpSet,
    cmpExact: cmpExact,
    cmpGrouped: cmpGrouped,
    compare: compare,
    signature: signature,
    COLUMNS: COLUMNS,
    LENGTH_CLOSE: LENGTH_CLOSE,
  };
}));
